_source/_posts/2022-07-21-angular-security-xss.md

```diff
@@ -0,0 +1,401 @@
+
+![Giphy of not sure if meme](https://media.giphy.com/media/cFgb5p5e1My3K/giphy.gif)
+
+Let's look at an example with a resource URL, such as the DOM-based XSS example where we bind the URL of the `iframe` source. 
+
+Your component code might look like this
+```ts
+@Component({
+  selector: 'app-video',
+  template: `
+    <iframe [src]="linky" width="800px" height="450px"
+  `
+})
+export class VideoComponent {
+
+  // pretend this is from an external source
+  public linky = '//videolink/embed/12345';
+}
+```
+
+Angular will stop you right there. 🛑
+
+You'll see an error in the console saying unsafe values can't be used in a resource URL. Angular recognizes you're trying to add a resource URL and is alarmed you're doing something dangerous. Resource URLs can contain legitimate code, so Angular can't sanitize it, unlike the comments we had above.
+
+If we are **sure** our link is safe and trustworthy (highly debatable in this example, but we'll ignore that for a moment), we can mark the resource as trusted after we do some cleanup to make the resource URL safer.
+
+
+> Remember, we should only use the `bypassSecurityTrust...` methods if we know the value is trustworthy. You bypass the built-in security mechanisms and open yourself to security vulnerabilities!
+
+Instead of using the entire video URL based on the external party's API response, we'll construct the URL by defining the video host URL within our app and appending the video's ID that we get back from the external party's API response. This way, we aren't entirely reliant on a potentially untrustworthy value from a third party. Instead, we'll have some measure of ensuring we're not injecting malicious code into the URL. 
+
+Then we'll mark the video URL as trusted and bind it in the template. Your `VideoComponent` changes to this:
+
+```ts
+@Component({
+  selector: 'app-video',
+  template: `
+    <iframe [src]="safeLinky" width="800px" height="450px"
+  `
+})
+export class VideoComponent {
+
+  // pretend this is from an external source
+  public videoId = '12345';
+  public safeLinky!: SafeResourceUrl;
+
+  constructor(private sanitizer: DomSanitizer) {
+    this.safeLinky = sanitizer.bypassSecurityTrustResourceUrl(`//videolink/embed/${this.videoId}`)
+  }
+}
+```
+
+Now you'll be able to show trailers of the K-Dramas on your site in an `iframe` in a much safer way.
+
+> It can't be stressed enough. Ensure you trust the values before bypassing the out-of-the-box security protections you get from Angular!
+
+Great! So we're done! Not quite. There are a couple of things to note.
+
+### Use ahead-of-time (AOT) compilation for extra security
+
+Angular's AOT compilation has extra security measures for injection attacks like XSS. AOT compilation is highly recommended for production code and has been the default compilation method since Angular v9. Not only is it more secure, but it also increases performance.
```

Comment:
```suggestion
Angular's AOT compilation has extra security measures for injection attacks like XSS. AOT compilation is highly recommended for production code and has been the default compilation method since Angular v9. Not only is it more secure, but it also improves performance.
```